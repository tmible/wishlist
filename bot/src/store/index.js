import migrate from '@tmible/wishlist-common/db-migrations';
import { provide } from '@tmible/wishlist-common/dependency-injector';
import Database from 'better-sqlite3';
import InjectionToken from '@tmible/wishlist-bot/architecture/injection-token';
import EditingModule from './editing/index.js';
import UsernamesModule from './usernames/index.js';
import WishlistModule from './wishlist/index.js';

/**
 * Обязательная часть элемента разметки текста
 * @typedef {object} EntityBase
 * @property {string} type Тип элемента
 * @property {number} offset Отступ от начала строки
 * @property {number} length Длина
 */
/**
 * Элемент разметки текста
 * @typedef {EntityBase & Record<string, unknown>} Entity
 */

/** @module Хранилище -- абстракция БД */

/**
 * Модули хранилища
 * @type {{ configure: Function } & Record<string, unknown>}
 */
const storeModules = [
  WishlistModule,
  EditingModule,
  UsernamesModule,
];

/**
 * Создание подключения к БД, [миграции БД]{@link migrate} и настройка модулей хранилища
 * @function initStore
 * @returns {() => void} Функция закрытия подключения к БД
 */
const initStore = () => {
  const db = new Database(process.env.WISHLIST_DB_FILE_PATH);
  migrate(db, process.env.WISHLIST_DB_MIGRATIONS_PATH);
  provide(InjectionToken.Database, db);
  storeModules.forEach(({ configure }) => configure());
  return () => db.close();
};

export default initStore;
