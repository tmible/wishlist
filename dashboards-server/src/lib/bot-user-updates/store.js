/** @typedef {import('./initialization.js').BotUserUpdatesDTO} BotUserUpdatesDTO */

/** @module Адаптер хранилища обновлений, полученных ботом */

/** @typedef {(value: any) => void} Subscription */

/**
 * Значение хранилица
 * @type {BotUserUpdatesDTO}
 */
const value = {
  page: [],
  total: 0,
  index: 0,
};

/**
 * Подписчики хранилища
 * @type {Map<Subscription, Subscription>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * обновлений, полученных ботом
 * @type {object}
 * @property {(subscription: Subscription) => () => void} subscribe Подписка на значение хранилища
 * @property {(BotUserUpdatesDTO) => void} set Установка значение хранилища
 */
export const botUserUpdates = {
  subscribe: (subsriber) => {
    subsriber(value);
    subscriptions.set(subsriber, subsriber);
    return () => subscriptions.delete(subsriber);
  },
  set: ({ page = [], total = 0, index = 0 }) => {
    value.page = page;
    value.total = total;
    value.index = index;
    for (const subsriber of subscriptions.values()) {
      subsriber(value);
    }
  },
};
