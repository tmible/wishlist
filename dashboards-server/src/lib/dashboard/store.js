/** @typedef {import('svelte/store').Readable} Readable */
/** @typedef {import('$lib/constants/period.const.js').PERIOD} PERIOD */
/** @typedef {import('./domain.js').Dashboard} Dashboard */
/** @typedef {import('./domain.js').DashboardData} DashboardData */

/** @module Адаптер хранилища дашборда */

/**
 * Хранилище дашборда
 * @typedef {object} Store
 * @implements {Readable<Dashboard>}
 * @property {(period: PERIOD) => void} updatePeriod Метод обновления периода
 * @property {(displayedCharts: { value: string }) => Promise<void>} updateChartsSelection
 *   Метод обновления выбранных графиков
 */

/**
 * Период автоматического обновления дашборда в милисекундах
 * @constant {number}
 */
const AUTO_UPDATE_INTERVAL = 5 * 60 * 1000;

/**
 * Создание хранилища дашборда
 * @param {Dashboard} initialValue Изначальные данные для графиков
 * @param {(chartKeys: string[], periodStart: number) => Promise<DashboardData<unknown>>} fetchData
 *   Функция получения данных для графиков
 * @returns {Store} Хранилище
 */
// eslint-disable-next-line max-lines-per-function -- много отдельных функций в общем контексте
export const createStore = (initialValue, fetchData) => {
  /**
   * Значение хранилища
   * @type {Dashboard}
   */
  const value = initialValue;

  /**
   * Подписчики хранилища
   * @template [S=(value: any) => void]
   * @type {Map<S, S>}
   * @see {@link https://svelte.dev/docs/svelte/stores#Store-contract}
   */
  const subscriptions = new Map();

  /**
   * Кэш для данных графиков с другим периодом. Сохраняет данные при выборе другого периода.
   * Очищается при очередном запросе данных в рамках автоматического обновления
   * @type {Map<PERIOD, [ string, number ]>}
   */
  const cache = new Map();

  /**
   * Таймаут запроса данных
   * @type {ReturnType<typeof setTimeout>}
   */
  let timeout;

  /**
   * Интервал запроса данных
   * @type {ReturnType<typeof setInterval>}
   */
  let interval;

  /* eslint-disable no-secrets/no-secrets -- просто длинное название */

  /**
   * Обновление данных графиков и оповещение подписчиков
   * @function updateChartsAndNotifySubscribers
   * @param {[ string, number ][]} data Пары идентификаторов и данных графиков
   * @returns {void}
   */
  /* eslint-enable no-secrets/no-secrets */
  const updateChartsAndNotifySubscribers = (data) => {
    for (const [ key, datum ] of data) {
      value.charts.get(key).data = datum;
    }
    for (const subscription of subscriptions.values()) {
      subscription(value);
    }
  };

  /**
   * [Получение данных для указанных графиков за актуальный период]{@link fetchData}.
   * Сохранение полученных значений, оповещение подпсчиков
   * @function fetcher
   * @param {string[]} keys Идентификаторы графиков, данные для которых нужно запросить.
   *   По умолчанию — все отображаемые графики
   * @returns {Promise<void>}
   * @async
   */
  const fetcher = async (
    keys = Array.from(value.charts.entries())
      .filter(([ , { isDisplayed }]) => isDisplayed)
      .map(([ key ]) => key),
  ) => {
    if (keys.length === 0) {
      return;
    }
    const data = await fetchData(keys, Date.now() - value.period);
    updateChartsAndNotifySubscribers(data.map((datum, i) => [ keys[i], datum ]));
  };

  /**
   * Инвалидация [кэша]{@link cache} и [обновление данных в хранилище]{@link fetcher}
   * @function updater
   * @param {unknown[]} args Проксируемые fetcher параметры
   * @returns {Promise<ReturnType<typeof fetcher>>} Возвращаемое из fetcher значение
   * @async
   */
  const updater = async (...args) => {
    cache.clear();
    return await fetcher(...args);
  };

  /**
   * Запуск или перезапуск автоматического обновления данных
   * @param {string[]} initialKeys Идентификаторы графиков для первого запроса данных
   * @returns {void}
   */
  const restartPolling = (initialKeys) => {
    if (timeout) {
      clearTimeout(timeout);
    }

    if (interval) {
      clearInterval(interval);
    }

    timeout = setTimeout(() => fetcher(initialKeys));
    interval = setInterval(updater, AUTO_UPDATE_INTERVAL);
  };

  /**
   * Сбор идентификаторов отображаемых графиков, для которых нет данных
   * @function collectNoDataCharts
   * @returns {string[]} Идентификаторы графиков
   */
  const collectNoDataCharts = () => {
    const noData = [];
    for (const [chartKey, chart] of value.charts.entries()) {
      if (chart.isDisplayed && chart.data === undefined) {
        noData.push(chartKey);
      }
    }
    return noData;
  };

  return {
    /**
     * Подписка на хранилище. Немедленный вызов функции-подписчика,
     * [сохранение подписки]{@link subscriptions}.
     * 1. При увеличении числа подписчиков с 0 до 1 — проверка наличия
     *    изначальных данных для графиков.
     * 1.1. При наличии изначальных данных — их установка, удаление изначальных данных,
     *      и [запуск автоматического обновления данных]{@link restartPolling}
     *      с первым запросом только за [недостающими данными]{@link collectNoDataCharts}.
     * 1.2. При отсутствии — [запуск автоматического обновления данных]{@link restartPolling}
     *      с параметрами по умолчанию.
     * 2. При уменьшении числа подписчиков с 1 до 0 — остановка автоматического обновления данных
     * @function subscribe
     * @param {(value: any) => void} subscription Функция-подписчик
     * @returns {() => void} Функция отписки
     * @see {@link https://svelte.dev/docs/svelte/stores#Store-contract}
     */
    subscribe: (subscription) => {
      subscription(value);
      subscriptions.set(subscription, subscription);

      if (subscriptions.size === 1) {
        if (Object.hasOwn(value, 'initialData')) {
          for (const [chartKey, chart] of value.charts.entries()) {
            if (Object.hasOwn(value.initialData, chartKey)) {
              chart.data = value.initialData[chartKey];
            }
          }
          restartPolling(collectNoDataCharts());
          delete value.initialData;
        } else {
          restartPolling();
        }
      }

      return () => {
        subscriptions.delete(subscription);

        if (subscriptions.size === 0) {
          clearTimeout(timeout);
          clearInterval(interval);
        }
      };
    },

    /**
     * Выбор нового периода для отображения графиков
     * 1. Сохранение в кэш данных для актуального преиода.
     * 2. Проверка наличия в кэше даннхы для выбравнного периода.
     * 2.1. При отсутствии — [перезапуск автоматического обновления данных]{@link restartPolling}.
     * 2.2. При наличии — установка значений из кэша и оповещение подписчиков.
     * @function updatePeriod
     * @param {PERIOD} period Новый выбранный период
     * @returns {void}
     */
    updatePeriod: (period) => {
      if (period && period !== value.period) {
        cache.set(
          value.period,
          Array.from(value.charts.entries(), ([ key, { data } ]) => [ key, data ]),
        );
        value.period = period;
      }
      const cached = cache.get(period);
      if (cached === undefined) {
        restartPolling();
      } else {
        updateChartsAndNotifySubscribers(cached);
      }
    },

    /**
     * Выбор графиков для отображения
     * Обновление признака отображения графиков,
     * [запрос]{@link fetcher} [недостающих данных]{@link collectNoDataCharts}
     * @function updateChartsSelection
     * @param {{ value: string }[]} displayedCharts
     *   Идентификаторы выбранных для отображения графиков
     * @returns {Promise<void>}
     * @async
     */
    updateChartsSelection: async (displayedCharts) => {
      for (const [chartKey, chart] of value.charts.entries()) {
        chart.isDisplayed = !!displayedCharts.some(({ value }) => value === chartKey);
      }
      await fetcher(collectNoDataCharts());
    },
  };
};
