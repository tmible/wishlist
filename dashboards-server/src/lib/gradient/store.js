/** @typedef {import('./domain.js').Gradient} Gradient */

/** @module Адаптер хранилища градиентов */

/**
 * Хранилище градиента
 * @template T
 * @typedef {object} Store
 * @property {() => T} get Получение градиента
 * @property {(gradient: T) => void} set Сохранение градиента
 * @property {() => void} delete Удаление градиента
 */

/**
 * Хранилище основного градиента
 * @type {Store<Gradient>}
 */
export const store = {
  get: () => JSON.parse(localStorage.getItem('gradient')),
  set: (gradient) => localStorage.setItem('gradient', JSON.stringify(gradient)),
  delete: () => localStorage.removeItem('gradient'),
};

/**
 * Следующий градиент
 * @type {Gradient}
 */
let nextGradient;

/**
 * Хранилище следующего градиента
 * @type {Store<Gradient>}
 */
export const nextStore = {
  get: () => nextGradient,
  set: (gradient) => nextGradient = gradient,
  delete: () => nextGradient = undefined,
};
