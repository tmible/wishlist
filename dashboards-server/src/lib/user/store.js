/** @module Адаптер хранилища пользователя */

/** @typedef {(value: any) => void} Subscription */
/** @typedef {{ isAuthenticated: boolean | null }} User */

/**
 * Значение хранилица
 * @type {User}
 */
let value = { isAuthenticated: null };

/**
 * Подписчики хранилища
 * @type {Map<Subscription, Subscription>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * пользователя
 * @type {object} Store
 * @property {(subscription: Subscription) => () => void} subscribe Подписка на значение хранилища
 * @property {(patchValue: Partial<User>) => void} patch Обновление значения хранилища
 */
export const user = {
  subscribe: (subscription) => {
    subscription(value);
    subscriptions.set(subscription, subscription);
    return () => subscriptions.delete(subscription);
  },
  patch: (patchValue) => {
    value = { ...value, ...patchValue };
    for (const subscription of subscriptions.values()) {
      subscription(value);
    }
  },
};
