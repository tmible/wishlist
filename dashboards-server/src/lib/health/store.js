import { readable } from 'svelte/store';
import { authInterceptor } from '$lib/auth-interceptor.js';
import { addOtherService } from './domain.js';

/** @typedef {import('svelte/store').Readable} Readable */
/** @typedef {import('./domain.js').HealthData} HealthData */

/**
 * Период автоматического обновления хранилища
 * @constant {number}
 */
const AUTO_UPDATE_PERIOD = 60 * 1000;

/**
 * Запрос данных о результатах последней проверки здоровья сервисов
 * @function fetchHealthData
 * @returns {Promise<HealthData>} Данные о результатах последней проверки здоровья сервисов
 * @async
 */
const fetchHealthData = async () => await fetch(
  '/api/data/health',
).then(
  authInterceptor,
).then(
  (response) => response.json(),
).then(
  addOtherService,
);

/**
 * Svelte хранилище данных о проверке здоровья сервисов
 * @constant {Readable<HealthData>}
 */
export const health = readable(
  {
    date: null,
    bot: null,
    portal: null,
    hub: null,
    refreshTokensCleaner: null,
    other: null,
  },
  (set) => {
    const timeout = setTimeout(async () => set(await fetchHealthData()));
    const interval = setInterval(async () => set(await fetchHealthData()), AUTO_UPDATE_PERIOD);
    return () => {
      clearTimeout(timeout);
      clearInterval(interval);
    };
  },
);
