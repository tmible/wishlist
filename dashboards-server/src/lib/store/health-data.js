import { readable } from 'svelte/store';
import { authInterceptor } from '$lib/auth-interceptor.js';

/** @typedef {import('svelte/store').Readable} Readable */
/**
 * Информация о здоровье бота
 * @typedef {object} BotHealthData
 * @property {boolean} service Запущен ли systemd сервис
 * @property {boolean} localhost Запущен ли http сервер
 * @property {boolean} https Проксируется ли https запрос к домену на http сервер
 * @property {boolean} dbConnection Подключен ли бот к БД
 * @property {boolean} localDBConnection Подключен ли бот к БД для горячих данных
 * @property {boolean} hubConnection Подключен ли бот к хабу
 */
/**
 * Информация о здоровье портала
 * @typedef {object} PortalHealthData
 * @property {boolean} service Запущен ли systemd сервис
 * @property {boolean} localhost Запущен ли http сервер
 * @property {boolean} https Проксируется ли https запрос к домену на http сервер
 * @property {boolean} dbConnection Подключен ли портал к БД
 * @property {boolean} hubConnection Подключен ли портал к хабу
 */
/**
 * Информация о здоровье хаба
 * @typedef {object} HubHealthData
 * @property {boolean} service Запущен ли systemd сервис
 * @property {boolean} socket Принимает ли хаб подключения к UNIX доменному сокету
 */
/**
 * Информация о проверке здоровья сервисов
 * @typedef {object} HealthData
 * @property {number} date Таймштамп проверки
 * @property {BotHealthData | null} bot Информация о здоровье бота
 * @property {PortalHealthData | null} portal Информация о здоровье портала
 * @property {HubHealthData | null} hub Информация о здоровье хаба
 */

/**
 * Запрос данных о результатах последней проверки здоровья сервисов
 * @function fetchHealthData
 * @returns {Promise<HealthData>} Данные о результатах последней проверки здоровья сервисов
 * @async
 */
const fetchHealthData = async () => await fetch(
  '/api/data/health',
).then(
  authInterceptor,
).then(
  (response) => response.json(),
);

/**
 * Svelte хранилище данных о проверке здоровья сервисов
 * @constant {Readable<HealthData>}
 */
export const healthData = readable(
  {
    date: null,
    bot: null,
    portal: null,
    hub: null,
  },
  (set) => {
    const timeout = setTimeout(async () => set(await fetchHealthData()));
    const interval = setInterval(async () => set(await fetchHealthData()), 60 * 1000);
    return () => {
      clearTimeout(timeout);
      clearInterval(interval);
    };
  },
);
