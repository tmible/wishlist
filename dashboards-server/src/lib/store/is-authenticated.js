import { writable } from 'svelte/store';

/** @typedef {import('svelte/store').Writable} Writable */

/**
 * Svelte хранилище признака аутентифицированности пользователя
 * @type {Writable<boolean | null>}
 */
export const isAuthenticated = writable(null);
