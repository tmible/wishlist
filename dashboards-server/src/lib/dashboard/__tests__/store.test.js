import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { createStore } from '../store.js';

describe('dashboard / store', () => {
  let initialValue;
  let fetchDataStub;

  beforeEach(() => {
    vi.spyOn(Date, 'now').mockReturnValue(10);
    initialValue = {
      period: 1,
      charts: new Map([
        [ 'chart 1', { data: undefined, isDisplayed: true } ],
        [ 'chart 2', { data: undefined, isDisplayed: true } ],
        [ 'chart 3', { data: undefined, isDisplayed: false } ],
      ]),
    };
    fetchDataStub = vi.fn(
      (keys, periodStart) => keys.map(
        (key) => `${key} data from network for period ${periodStart}`,
      ),
    );
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('subscribe', () => {
    it('should immediately call subscription', () => {
      const subscriber = vi.fn();
      const unsubscribe = createStore('initial value', vi.fn()).subscribe(subscriber);
      unsubscribe();
      expect(subscriber).toHaveBeenCalledWith('initial value');
    });

    describe('on first subscriber', () => {
      beforeEach(() => {
        vi.stubGlobal('setInterval', vi.fn());
      });

      describe('if there is initial data', () => {
        let store;
        let unsubscribe;

        beforeEach(() => {
          initialValue.initialData = { 'chart 1': 'chart 1 initial data' };
          store = createStore(initialValue, fetchDataStub);
          vi.useFakeTimers();
        });

        afterEach(() => {
          unsubscribe();
          vi.useRealTimers();
          delete initialValue.initialData;
        });

        it('should set initial data and fetch missing data', async () => {
          let value;
          unsubscribe = store.subscribe(vi.fn((storeValue) => value = storeValue));
          await vi.advanceTimersToNextTimerAsync();
          expect(value).toMatchSnapshot();
        });

        it('should not reuse initial data', async () => {
          unsubscribe = store.subscribe(vi.fn());
          await vi.advanceTimersToNextTimerAsync();
          unsubscribe();
          let value;
          unsubscribe = store.subscribe(vi.fn((storeValue) => value = storeValue));
          await vi.advanceTimersToNextTimerAsync();
          expect(value).toMatchSnapshot();
        });
      });

      describe('if there is no initial data', () => {
        it('should fetch all charts data', async () => {
          let value;
          vi.useFakeTimers();
          const unsubscribe = createStore(
            initialValue,
            fetchDataStub,
          ).subscribe(
            vi.fn((storeValue) => value = storeValue),
          );
          await vi.advanceTimersToNextTimerAsync();
          unsubscribe();
          vi.useRealTimers();
          expect(value).toMatchSnapshot();
        });
      });

      it('should set auto update interval', () => {
        const unsubscribe = createStore('initial value', vi.fn()).subscribe(vi.fn());
        unsubscribe();
        expect(vi.mocked(setInterval)).toHaveBeenCalled();
      });
    });

    describe('on last subscriber', () => {
      it('should clear timeout', () => {
        vi.stubGlobal('setTimeout', vi.fn().mockReturnValueOnce('timeout'));
        vi.stubGlobal('clearTimeout', vi.fn());
        createStore('initial value', vi.fn()).subscribe(vi.fn())();
        expect(vi.mocked(clearTimeout)).toHaveBeenCalledWith('timeout');
      });

      it('should clear interval', () => {
        vi.stubGlobal('setInterval', vi.fn().mockReturnValueOnce('interval'));
        vi.stubGlobal('clearInterval', vi.fn());
        createStore('initial value', vi.fn()).subscribe(vi.fn())();
        expect(vi.mocked(clearInterval)).toHaveBeenCalledWith('interval');
      });
    });
  });

  describe('updatePeriod', () => {
    it('should fetch data if it wasn\'t used since last auto update', async () => {
      let value;
      const store = createStore(initialValue, fetchDataStub);
      vi.useFakeTimers();
      vi.stubGlobal('setInterval', vi.fn());
      const unsubscribe = store.subscribe(vi.fn((storeValue) => value = storeValue));
      await vi.advanceTimersToNextTimerAsync();
      store.updatePeriod(2);
      await vi.advanceTimersToNextTimerAsync();
      unsubscribe();
      vi.useRealTimers();
      expect(value).toMatchSnapshot();
    });

    it('should not fetch data if it was used since last auto update', async () => {
      const store = createStore(initialValue, fetchDataStub);
      vi.useFakeTimers();
      vi.stubGlobal('setInterval', vi.fn());
      const unsubscribe = store.subscribe(vi.fn());
      await vi.advanceTimersToNextTimerAsync();
      store.updatePeriod(2);
      await vi.advanceTimersToNextTimerAsync();
      fetchDataStub.mockClear();
      store.updatePeriod(1);
      await vi.advanceTimersToNextTimerAsync();
      unsubscribe();
      vi.useRealTimers();
      expect(fetchDataStub).not.toHaveBeenCalled();
    });
  });

  describe('updateChartsSelection', () => {
    it('should fetch missing data', async () => {
      let value;
      const store = createStore(initialValue, fetchDataStub);
      vi.useFakeTimers();
      vi.stubGlobal('setInterval', vi.fn());
      const unsubscribe = store.subscribe(vi.fn((storeValue) => value = storeValue));
      await vi.advanceTimersToNextTimerAsync();
      store.updateChartsSelection([ 'chart 1', 'chart 3' ]);
      await vi.advanceTimersToNextTimerAsync();
      unsubscribe();
      vi.useRealTimers();
      expect(value).toMatchSnapshot();
    });

    it('should notify subscribers even though no data is missing', async () => {
      let value;
      const store = createStore(initialValue, fetchDataStub);
      vi.useFakeTimers();
      vi.stubGlobal('setInterval', vi.fn());
      const unsubscribe = store.subscribe(vi.fn((storeValue) => value = storeValue));
      await vi.advanceTimersToNextTimerAsync();
      store.updateChartsSelection([ 'chart 1' ]);
      await vi.advanceTimersToNextTimerAsync();
      unsubscribe();
      vi.useRealTimers();
      expect(value).toMatchSnapshot();
    });
  });

  describe('auto update', () => {
    it('should auto update displayed charts', async () => {
      initialValue.initialData = { 'chart 1': 'chart 1 initial data' };
      let value;
      const store = createStore(initialValue, fetchDataStub);
      vi.useFakeTimers();
      vi.spyOn(Date, 'now').mockReturnValue(10);
      const unsubscribe = store.subscribe(vi.fn((storeValue) => value = storeValue));
      await vi.advanceTimersToNextTimerAsync();
      await vi.advanceTimersToNextTimerAsync();
      unsubscribe();
      vi.useRealTimers();
      expect(value).toMatchSnapshot();
    });

    it('should clear periods cache', async () => {
      const store = createStore(initialValue, fetchDataStub);
      vi.useFakeTimers();
      vi.spyOn(Date, 'now').mockReturnValue(10);
      const unsubscribe = store.subscribe(vi.fn());
      await vi.advanceTimersToNextTimerAsync();
      store.updatePeriod(2);
      await vi.advanceTimersToNextTimerAsync();
      await vi.advanceTimersToNextTimerAsync();
      fetchDataStub.mockClear();
      store.updatePeriod(1);
      await vi.advanceTimersToNextTimerAsync();
      unsubscribe();
      vi.useRealTimers();
      expect(fetchDataStub).toHaveBeenCalled();
    });
  });
});
