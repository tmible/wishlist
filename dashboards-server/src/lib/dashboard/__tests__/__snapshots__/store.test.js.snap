// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dashboard / store > auto update > should auto update displayed charts 1`] = `
{
  "charts": Map {
    "chart 1" => {
      "data": "chart 1 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 2" => {
      "data": "chart 2 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 3" => {
      "data": undefined,
      "isDisplayed": false,
    },
  },
  "period": 1,
}
`;

exports[`dashboard / store > subscribe > on first subscriber > if there is initial data > should not reuse initial data 1`] = `
{
  "charts": Map {
    "chart 1" => {
      "data": "chart 1 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 2" => {
      "data": "chart 2 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 3" => {
      "data": undefined,
      "isDisplayed": false,
    },
  },
  "period": 1,
}
`;

exports[`dashboard / store > subscribe > on first subscriber > if there is initial data > should set initial data and fetch missing data 1`] = `
{
  "charts": Map {
    "chart 1" => {
      "data": "chart 1 initial data",
      "isDisplayed": true,
    },
    "chart 2" => {
      "data": "chart 2 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 3" => {
      "data": undefined,
      "isDisplayed": false,
    },
  },
  "period": 1,
}
`;

exports[`dashboard / store > subscribe > on first subscriber > if there is no initial data > should fetch all charts data 1`] = `
{
  "charts": Map {
    "chart 1" => {
      "data": "chart 1 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 2" => {
      "data": "chart 2 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 3" => {
      "data": undefined,
      "isDisplayed": false,
    },
  },
  "period": 1,
}
`;

exports[`dashboard / store > updateChartsSelection > should fetch missing data 1`] = `
{
  "charts": Map {
    "chart 1" => {
      "data": "chart 1 data from network for period 9",
      "isDisplayed": true,
    },
    "chart 2" => {
      "data": "chart 2 data from network for period 9",
      "isDisplayed": false,
    },
    "chart 3" => {
      "data": "chart 3 data from network for period 9",
      "isDisplayed": true,
    },
  },
  "period": 1,
}
`;

exports[`dashboard / store > updatePeriod > should fetch data if it wasn't used since last auto update 1`] = `
{
  "charts": Map {
    "chart 1" => {
      "data": "chart 1 data from network for period 8",
      "isDisplayed": true,
    },
    "chart 2" => {
      "data": "chart 2 data from network for period 8",
      "isDisplayed": true,
    },
    "chart 3" => {
      "data": undefined,
      "isDisplayed": false,
    },
  },
  "period": 2,
}
`;
